### ISSUE TITLE
A regex rule in a filter list leads to AdGuard hanging the browser

### ISSUE THREAD

**Issue by Xiarno** [AdguardTeam/AdguardBrowserExtension#2240]
# IMPORTANT: the quick fix would be to check the filters updates and restart the browser afterwards.

### AdGuard Extension version

4.1.46

### Browser version

Chrome 108.0.5359.99

### OS version

Windows 10 Pro 10.0.19044 Build 19044

### What filters do you have enabled?

AdGuard Base filter, AdGuard Chinese filter, AdGuard Dutch filter, AdGuard French filter, AdGuard German filter, AdGuard Japanese filter, AdGuard Russian filter, AdGuard Spanish/Portuguese filter, AdGuard Turkish filter, AdGuard Tracking Protection filter, Filter unblocking search ads and self-promotion, ABPindo, EasyList China, EasyList Czech and Slovak, EasyList Italy, Official Polish filters for AdBlock, uBlock Origin & AdGuard, List-KR, Adblock List for Finland, NoCoin Filter List, Frellwit's Swedish Filter, Dandelion Sprout's Nordic Filters

### Issue Details

Suddenly, all webpages stopped loading. Even after restarting the browser or even the computer, it would persist. The only way to make it go back to normal was to disable the extension from Chrome completely. Re-enabling it immediately resumes the problem.

I have multiple other chrome accounts that do not have this problem occurring. It really only started occurring on one suddenly.

### Expected Behavior

A normal working browser with the extension enable.

**Comment by StruffelProductions** [comment]
The same is happening for me. Also version 4.1.46.

**Comment by ngotrongphuc** [comment]
i got the same issue, chrome suddenly got infinite loading even though other browsers are working fine

**Comment by RoachCoach521** [comment]
Chiming in - same problem. Was pulling my hair out with this, started disabling extensions one by one - Adguard was the issue.

Disabled, everything works perfectly.

**Comment by code86** [comment]
I have the same problem. Glad I am not alone.

**Comment by lunamoth** [comment]
I also had that problem.

**Comment by stx** [comment]
Same problem.

@slavaleleka @zhelvis and others: FYI, you're having an emergency here.

**Comment by MscXander81** [comment]
Same thing for me for 2 hours. Initially I thought it was PC or Internet and instead I discover by starting Chrome with incognito browsing that it was the extensions that blocked the whole browser and restarting them one at a time discovered that it was **AdGuard AdBlocker 4.1.46** which prevented the loading of any web page (Italian user)

**Comment by maximtop** [comment]
I can't reproduce it on my side. Looks like the filter issue or browser issue.
Have you tried to disable filters one by one? Or reinstall the extension? Or install the extension in canary?

**Comment by maximtop** [comment]
I was able to reproduce it. It looks like the problem is in the List KR filter, if you do not enable it, the extension works.

<img width="740" alt="image" src="https://user-images.githubusercontent.com/17272769/206796975-02b245ad-3ad9-467e-a9ec-f9e6aa23039c.png">

**Comment by Xiarno** [comment]
> I was able to reproduce it. It looks like the problem is in the List KR filter, if you do not enable it, the extension works.
> 
> <img alt="image" width="740" src="https://user-images.githubusercontent.com/17272769/206796975-02b245ad-3ad9-467e-a9ec-f9e6aa23039c.png">

Disabled it. Seems fine for now! Awesome!

**Comment by RoachCoach521** [comment]
The dumb problem I'm having is that when I enable, it causes infinite hangs in any tab I try to access its settings from, so I'm unable to actually disable the list.

**Comment by maximtop** [comment]
The rule causing the problem. 
```
/^http(s|):\/\/([a-z0-9-\.]+|)+[a-z0-9-]+\.[a-z]+\/adManager\/(css|js)\/[A-z]+\.(css|js)$/
```
We will investigate why this rule hangs out the extension.

**Comment by bradtem** [comment]
For those disabling it, you need to be quick.  You can't toggle the lists while the extension is disabled.   But turn it on and you'll have it pinning the CPU soon and you can't disable it.    So quickly go in, disable all the language extensions, then go in to disable list-kr, and re-enable the rest.  Seemed to work.  Though presumably an update will come soon enough.

I wondered if this was some web site who found a way to make the ad-blocker pin the CPU in order to get me to disable my ad-blocker.  I did disable it once I saw it was the problem, that's for sure.

**Comment by Alex-302** [comment]
If you cannot chech for filters updates, please
1. disable/enable AdGuard in browser settings
2. click on AdGuard icon
3. click `Check for updates` button.
4. 
Do not open settings or sites after disable/enable an extension.

<details>
<summary>Details</summary>

![image](https://user-images.githubusercontent.com/8361299/206809070-37d84689-63c1-4f3d-b054-0d72b36fef63.png)


</details>

**Comment by ameshkov** [comment]
Thank you so much for reporting it! This is a major bug and we'll be rolling a patch that resolves this asap.

**Comment by al2o3cr** [comment]
The nested `+` with a zero-length alternative drives the engine to backtrack heavily, especially with a non-matching character in the string. For instance, detecting that this doesn't match takes 2.5ms (with [Regexr](https://regexr.com/)):
```
https://zzz.foo123.ba!r.baz/adManager/css/ABC.css
```
this one takes 5.3ms:
```
https://zzz.foo123a.ba!r.baz/adManager/css/ABC.css
```
Adding a couple more `a`s after `123` pushes the runtime to **129.4ms**:
```
https://zzz.foo123aaaaaa.ba!r.baz/adManager/css/ABC.css
```
Execution time is roughly _doubling_ per character added that could match in `([a-z0-9-\.]+|)+`

The outer `+` only ever matches one time in successful cases, so I believe that using `([a-z0-9-\.]+)?` will accomplish the same goal without the undesired backtracking behavior.

**Comment by ameshkov** [comment]
A slow regex is one thing, but here we're dealing with some sort of an infinite loop inside Chrome's RegExp, once it's stuck, it never stops and consumes 100% cpu. It happens only with some input strings, not with all of them. I'll try to narrow it down to something more specific tomorrow.

**Comment by master4x** [comment]
My Chrome browser catched the following error during testing:

`pages/content-script-start.js:2662 (anonymous function)`
```
const wrappedSendMessage = (name, metadata, apiNamespaceObj, ...args) => {
        if (args.length < metadata.minArgs) {
          throw new Error(`Expected at least ${metadata.minArgs} ${pluralizeArguments(metadata.minArgs)} for ${name}(), got ${args.length}`);
        }

        if (args.length > metadata.maxArgs) {
          throw new Error(`Expected at most ${metadata.maxArgs} ${pluralizeArguments(metadata.maxArgs)} for ${name}(), got ${args.length}`);
        }

        return new Promise((resolve, reject) => {
          const wrappedCb = wrappedSendMessageCallback.bind(null, {
            resolve,
            reject
          });
          args.push(wrappedCb);
          apiNamespaceObj.sendMessage(...args);
        });
      };
```

**Comment by al2o3cr** [comment]
@ameshkov "infinite" and "exponential" are hard to tell apart - based on the scaling in my previous post, trying to match this URL which has 34 more `a`s would take about **1000 years**:
```
https://zzz.foo123aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.ba!r.baz/adManager/css/ABC.css
```

**Comment by gorhill** [comment]
> I disabled the rule temporary until this bug is resolved

@piquark6046 The regex suffers from "catastrophic backtracking"[1] as pointed out by @al2o3cr.

I see that the filter is still enabled for uBO, which could suffer the same issue as AdGuard if it was not for the fact that in uBO the regex is executed only when the keyword `admanag` appears in a URL. I suggest you rewrite the regex to avoid empty string alternatives:

    /^https?:\/\/([a-z0-9-\.]+)?[a-z0-9-]+\.[a-z]+\/adManager\/(css|js)\/[A-z]+\.(css|js)$/

---

[1] See <https://regex101.com/r/8tj50E/1>

**Comment by piquark6046** [comment]
@gorhill I will push a commit containing the following filters:
```adblock
/^http(s|):\/\/[a-z0-9-\.]*[a-z0-9-]+\.[a-z]+\/bannerpop\/uploads\//
/^http(s|):\/\/[a-z0-9-\.]*[a-z0-9-]+\.[a-z]+\/adManager\/(css|js)\/[A-z]+\.(css|js)$/
```

**Comment by gorhill** [comment]
Why use `http(s|):` when `https?:` does the same and without potentially bad side-effect and semantically better? Using empty string alternative should never be needed, regex syntax has all the proper special character to deal with optional parts.

**Comment by piquark6046** [comment]
Okay. I will apply your opinion.

**Comment by gorhill** [comment]
`?` is a quantifier, it means "zero or one".

**Comment by ameshkov** [comment]
@al2o3cr you're right. I suppose trying to introduce protection from ReDos inside the extension itself might be too problematic.

We'll consider adding protection against ReDos on the stage when filter lists are being compiled server-side:
https://github.com/AdguardTeam/FiltersCompiler/issues/162

**Comment by krystian3w** [comment]
> I wondered if this was some web site who found a way to make the ad-blocker pin the CPU in order to get me to disable my ad-blocker. I did disable it once I saw it was the problem, that's for sure.

`wp.pl` - if ads or video player back then people open issues about CPU/RAM abuse.

**Comment by fedor7** [comment]
After update Adguard to 4.1.52, firefox is not working. CPU 100%

**Comment by maximtop** [comment]
> After update Adguard to 4.1.52, firefox is not working. CPU 100%

I've checked briefly and it seems the extension is working as expected. Please feel free to open a new issue with information about filters enabled.
